include/instantiate.h

```diff
@@ -0,0 +1,168 @@
+    }
+  };
+
+  /**
+     @brief Instantiate the reconstruction template at index i and
+     recurse to prior element
+  */
+  template <template <typename, int, QudaReconstructType> class Apply, typename Float, int nColor, typename Recon,
+            int i, typename G, typename... Args>
+  struct instantiateReconstruct {
+    instantiateReconstruct(G &U, Args &&... args)
+    {
+      if (U.Reconstruct() == Recon::recon[i]) {
+        instantiateApply<is_enabled<Recon::recon[i]>(), Apply, Float, nColor, Recon::recon[i], G, Args...>(U, args...);
+      } else {
+        instantiateReconstruct<Apply, Float, nColor, Recon, i - 1, G, Args...>(U, args...);
+      }
+    }
+  };
+
+  /**
+     @brief Termination specialization of instantiateReconstruct
+  */
+  template <template <typename, int, QudaReconstructType> class Apply, typename Float, int nColor, typename Recon,
+            typename G, typename... Args>
+  struct instantiateReconstruct<Apply, Float, nColor, Recon, 0, G, Args...> {
+    instantiateReconstruct(G &U, Args &&... args)
+    {
+      if (U.Reconstruct() == Recon::recon[0]) {
+        instantiateApply<is_enabled<Recon::recon[0]>(), Apply, Float, nColor, Recon::recon[0], G, Args...>(U, args...);
+      } else {
+        errorQuda("Unsupported reconstruct type %d\n", U.Reconstruct());
+      }
+    }
+  };
+
+  /**
+     @brief This instantiate function is used to instantiate the colors
+     @param[in] U Gauge field
+     @param[in,out] args Additional arguments for kernels
+  */
+  template <template <typename, int, QudaReconstructType> class Apply, typename Recon, typename Float, typename G,
+            typename... Args>
+  constexpr void instantiate(G &U, Args &&... args)
+  {
+    if (U.Ncolor() == 3) {
+      constexpr int i = Recon::recon.size() - 1;
+      instantiateReconstruct<Apply, Float, 3, Recon, i, G, Args...>(U, args...);
+    } else {
+      errorQuda("Unsupported number of colors %d\n", U.Ncolor());
+    }
+  }
+
+  /**
+     @brief This instantiate function is used to instantiate the precisions
+     @param[in] U Gauge field
+     @param[in,out] args Additional arguments for different dslash kernels
+  */
+  template <template <typename, int, QudaReconstructType> class Apply, typename Recon = ReconstructFull, typename G,
+            typename... Args>
+  constexpr void instantiate(G &U, Args &&... args)
+  {
+    if (U.Precision() == QUDA_DOUBLE_PRECISION) {
+#if QUDA_PRECISION & 8
+      instantiate<Apply, Recon, double>(U, args...);
+#else
+      errorQuda("QUDA_PRECISION=%d does not enable double precision", QUDA_PRECISION);
+#endif
+    } else if (U.Precision() == QUDA_SINGLE_PRECISION) {
+#if QUDA_PRECISION & 4
+      instantiate<Apply, Recon, float>(U, args...);
+#else
+      errorQuda("QUDA_PRECISION=%d does not enable single precision", QUDA_PRECISION);
+#endif
+    } else {
+      errorQuda("Unsupported precision %d\n", U.Precision());
```

Comment:
Why does this terminate after `single`? Alternatively, am I slightly missing where this is used, and the doxygen of `Additional arguments for different dslash kernels` isn't correct?

Reply:
This helper is only used for gauge algorithms so we don’t need to have half or quarter precision support in there.

I will fix the doxgyen

Reply:
👍